core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -218,6 +218,7 @@ private[spark] class Executor(
           threwException = false
           res
         } finally {
+          env.blockManager.releaseAllPinsForTask(taskId)
```

Comment:
Can we verify in the non-failure case that the pins are 0? Maybe log a warning or something. Similar to how freedMemory is working below.